

#' plot_ele_depth
#'
#' Plot the depth of elements.
#'
#' @param depth_data Depth data generated by the APAV tool.
#' @param ele_data Element data generated by the APAV tool.
#' @param gff_data GFF data.
#' @param pheno_data Phenotype data. The row names are sample names.
#'
#' @param cluster_samples A logical value indicating whether perform clustering on samples.
#' @param clustering_distance Method of measuring distance when clustering, pass to \code{\link[stats]{dist}}.
#' @param clustering_method Method to perform hierarchical clustering, pass to \code{\link[stats]{hclust}}.
#'
#' @param log10 A logical value indicating whether convert depth value to log10(depth+1).
#' @param ele_region Gene element regions for display, choosing from "TN"(transcripts number), "UP"(upstream) and "DOWN"(downstream).
#'
#' @param top_anno_height The relative height of annotation, ranging from 0 to 1.
#' @param left_anno_width The relative width of annotation, ranging from 0 to 1.
#'
#' @param ele_color The color of element regions.
#' @param depth_colors A vector of colors of depth.
#' @param lowlight_colors A vector of colors for depth of non-element region.
#' Element region not highlighted when set to NULL.
#' @param gene_colors A named vector for gene elements.
#' e.g. c(CDS = "#d77478", five_prime_UTR = "#ecbf40")
#' @param pheno_info_color_list A list contains named vector of colors for phenotype.
#' e.g. list(gender = c("Male" = "green", "Female" = "red"), age = c("yellow", "red"))
#' @param pheno_border The color of cell borders in phenotype annotation.
#'
#' @param seq_name_size The size of sequence name.
#' @param loc_name_size The size of coordinate numbers.
#'
#' @param show_sample_name A logical value indicating whether show sample names.
#' @param sample_name_size The size of sample names.
#'
#' @param legend_title The title of legend.
#' @param legend_title_size The size of legend title.
#' @param legend_text_size The size of legend item labels.
#'
#' @import data.table
#' @import ggplot2
#' @import patchwork
#' @importFrom ggnewscale new_scale_fill
#'
#' @export
#'

plot_ele_depth <- function(
    depth_data,
    ele_data,
    gff_data = NULL,
    pheno_data = NULL,

    cluster_samples = F,
    clustering_distance = "euclidean",
    clustering_method = "complete",

    log10 = F,
    ele_region = NULL,

    top_anno_height = 0.4,
    left_anno_width = 0.1,

    ele_color = "#A6CEE3",
    depth_colors = c("#5680ae", "white"),
    lowlight_colors = c("gray", "white"),
    gene_colors = structure(c("gray70", "gray85", "#d77478", "#ecbf40", "#ecbf40"),
                            names = c("transcript", "exon", "CDS", "five_prime_UTR", "three_prime_UTR")),
    pheno_info_color_list = NULL,
    pheno_border = NA,

    seq_name_size = NULL,
    loc_name_size = NULL,

    show_sample_name = T,
    sample_name_size = NULL,

    legend_title = "Depth",
    legend_title_size = NULL,
    legend_text_size = NULL
  ){

  match_color("ele_color", "#A6CEE3", ele_color, 1)
  match_color("depth_colors", c("#5680ae", "white"), depth_colors)
  if(!is.null(depth_colors)){
    match_color("lowlight_colors", c("gray", "white"), lowlight_colors)
  }
  gene_colors <- get_gene_palette(gene_colors)

  if(!is.null(pheno_data)){
    pheno_colors <- get_anno_palette(pheno_info_color_list, as.list(pheno_data))
  }else{
    pheno_colors <- NULL
  }

  top_anno_height <- match_num("top_anno_height", top_anno_height, 0, 1)
  left_anno_width <- match_num("left_anno_width", left_anno_width, 0, 1)

  seqname <- depth_data$Chr[1]

  loc_start <- min(depth_data$Pos)
  loc_end <- max(depth_data$Pos)

  if(!is.null(gff_data)){
    colnames(gff_data) <- c("seqid", "source", "type", "start", "end", "score", "strand", "phase", "attr")
    loc_start <- min(loc_start, min(gff_data$start))
    loc_end <- max(loc_end, max(gff_data$end))
  }

  if(!is.null(ele_data)){
    if(is.numeric(ele_data$Start)){
      ele_data <- ele_data[order(ele_data$Start),]
      loc_start <- min(loc_start, min(ele_data$Start))
      loc_end <- max(loc_end, max(ele_data$End))
    }else{
      ele_data <- ele_data[order(as.numeric(unlist(lapply(strsplit(ele_data$Start, ";|,"), function(x){x[[1]]})))),]
      loc_start <- min(loc_start, min(as.numeric(unlist(strsplit(ele_data$Start, ";|,")))))
      loc_end <- max(loc_end, max(as.numeric(unlist(strsplit(ele_data$End, ";|,")))))
    }
  }

  loc_start <- loc_start - 1
  loc_end <- loc_end + 1

  if(!is.null(ele_data)){
    if(length(grep(";",c(ele_data$Start, ele_data$End))) > 0){
      for(i in 1:nrow(ele_data)){
        pos_m <- merge_block_pos(ele_data[i, "Start"], ele_data[i, "End"])
        ele_data[i, "Start"] <- pos_m[1]
        ele_data[i, "End"] <- pos_m[2]
        ele_data[i, "Length"] <- pos_m[3]
      }
    }

    if(!is.null(ele_region)){
      fdata <- filter_eledata(ele_data, ele_region)
      if(is.null(fdata)){
        stop("Can not find ",ele_region,".\n")
      }
    }else{
      fdata <- ele_data
    }

    p_ele_res <- get_ele(fdata, ele_color, NA, loc_start, loc_end)
    p_ele <- p_ele_res$p
    enrow <- p_ele_res$nrow
    ele_pos <- get_blocks_region(fdata$Start, fdata$End)
  }else{
    lowlight_colors <- NULL
    ele_pos <- NULL
  }

  if(log10){
    legend_title <- "log10(Depth+1)"
  }

  ps_depth_res <- get_depth_ps(depth_data,
                               cluster_samples,
                               clustering_distance,
                               clustering_method,
                               log10,
                               depth_colors,
                               pheno_data,
                               pheno_colors,
                               pheno_border,
                               lowlight_colors,
                               ele_pos,
                               show_sample_name,
                               sample_name_size,
                               legend_title,
                               legend_title_size,
                               legend_text_size
                               )
  p_depth <- ps_depth_res$depth
  p_anno <- ps_depth_res$anno


  if(!is.null(gff_data)){
    p_struc_res <- get_struc(gff_data, gene_colors)
    p_struc <- p_struc_res$p
    gnrow <- p_struc_res$nrow
  }


  if(is.null(ele_data)){
    if(is.null(gff_data)){
      p_depth <- add_coord(p_depth, seqname, loc_start, loc_end, seq_name_size, loc_name_size)
      p_anno + p_depth + plot_layout(widths = c(left_anno_width, 1 - left_anno_width))
    }else{
      p_struc <- add_coord(p_struc, seqname, loc_start, loc_end, seq_name_size, loc_name_size)
      p_depth <- p_depth +
        scale_x_continuous(expand = expansion(mult=0), limits = c(loc_start, loc_end))
      p_struc + p_depth + p_anno +
        patchwork::plot_layout(design = "
                      #1
                      32
                      ",
                               heights = c(top_anno_height, 1 - top_anno_height),
                               widths= c(left_anno_width, 1 - left_anno_width))
    }
  }else{
    if(is.null(gff_data)){
      p_ele <- add_coord(p_ele, seqname, loc_start, loc_end, seq_name_size, loc_name_size)
      p_depth <- p_depth +
        scale_x_continuous(expand = expansion(mult=0), limits = c(loc_start, loc_end))
      p_ele + p_depth + p_anno +
        patchwork::plot_layout(design = "
                      #1
                      32
                      ",
                               heights = c(top_anno_height, 1 - top_anno_height),
                               widths= c(left_anno_width, 1 - left_anno_width))
    }else{
      p_struc <- add_coord(p_struc, seqname, loc_start, loc_end, seq_name_size, loc_name_size)
      p_ele <- p_ele +
        scale_x_continuous(expand = expansion(mult=0), limits = c(loc_start, loc_end))
      p_depth <- p_depth +
        scale_x_continuous(expand = expansion(mult=0), limits = c(loc_start, loc_end))
      p_struc + p_ele + p_depth + p_anno +
        patchwork::plot_layout(design = "
                      #1
                      #2
                      43
                      ",
                               heights = c(top_anno_height * (gnrow / (gnrow + enrow)),
                                           top_anno_height * (enrow / (gnrow + enrow)),
                                           (1 - top_anno_height)),
                               widths= c(left_anno_width, 1 - left_anno_width))
    }
  }



}


#' plot_ele_cov
#'
#' Plot the coverage of elements.
#'
#' @param data Coverage results data generated by the APAV tool.
#' @param gff_data GFF data.
#' @param pheno_data Phenotype data. The row names are sample names.
#'
#' @param cluster_samples A logical value indicating whether perform clustering on samples.
#' @param clustering_distance Method of measuring distance when clustering, pass to \code{\link[stats]{dist}}.
#' @param clustering_method Method to perform hierarchical clustering, pass to \code{\link[stats]{hclust}}.
#'
#' @param ele_region Gene element regions for display, choosing from "TN"(transcripts number), "UP"(upstream) and "DOWN"(downstream).
#'
#' @param top_anno_height The relative height of annotation, ranging from 0 to 1.
#' @param left_anno_width The relative width of annotation, ranging from 0 to 1.
#'
#' @param ele_color The color of element regions.
#' @param ele_line_color The color for the line pointing to elements.
#' @param cov_colors A vector of colors for coverage.
#' @param cell_border The color of cell borders in heat map.
#' @param gene_colors A named vector for gene elements.
#' e.g. c(CDS = "#d77478", five_prime_UTR = "#ecbf40")
#' @param pheno_info_color_list A list contains named vector of colors for phenotype.
#' e.g. list(gender = c("Male" = "green", "Female" = "red"), age = c("yellow", "red"))
#' @param pheno_border The color of cell borders in phenotype annotation.
#'
#' @param seq_name_size The size of sequence name.
#' @param loc_name_size The size of coordinate numbers.
#'
#' @param show_ele_name A logical value indicating whether show element names.
#' @param ele_name_size The size of element names.
#' @param ele_name_len The longest length of element names.
#' @param ele_name_rot The rotation of element names.
#'
#' @param show_sample_name A logical value indicating whether show sample names.
#' @param sample_name_size The size of sample names.
#'
#' @param legend_title_size The size of legend title.
#' @param legend_text_size The size of legend item labels.
#'
#' @import data.table
#' @import ggplot2
#' @import patchwork
#' @importFrom ggnewscale new_scale_fill
#'
#' @export
#'

plot_ele_cov <- function(
    data,
    gff_data = NULL,
    pheno_data = NULL,

    cluster_samples = F,
    clustering_distance = "euclidean",
    clustering_method = "complete",

    ele_region = NULL,

    top_anno_height = 0.4,
    left_anno_width = 0.1,

    ele_color = "#A6CEE3",
    ele_line_color = "gray",
    cov_colors = c("#5680ae", "gray90"),
    cell_border = NA,
    gene_colors = structure(c("gray70", "gray85", "#d77478", "#ecbf40", "#ecbf40"),
                            names = c("transcript", "exon", "CDS", "five_prime_UTR", "three_prime_UTR")),
    pheno_info_color_list = NULL,
    pheno_border = NA,

    seq_name_size = NULL,
    loc_name_size = NULL,

    show_ele_name = T,
    ele_name_size = NULL,
    ele_name_rot = 90,
    ele_name_len = 20,

    show_sample_name = T,
    sample_name_size = NULL,

    legend_title_size = NULL,
    legend_text_size = NULL
  ){

  if(length(grep(";",c(data$Start, data$End))) > 0){
    stop("Please check the input data. It looks like merged PAV profile instead of coverage profile.")
  }

  match_color("cov_colors", c("#5680ae", "gray90"), cov_colors)

  plot_ele(
    "Coverage",
    data,
    ele_region,
    gff_data,
    pheno_data,
    cluster_samples,
    clustering_distance,
    clustering_method,
    top_anno_height,
    left_anno_width,
    ele_color,
    ele_line_color,
    cov_colors,
    cell_border,
    gene_colors,
    pheno_info_color_list,
    pheno_border,
    seq_name_size,
    loc_name_size,
    show_ele_name,
    ele_name_size,
    ele_name_rot,
    ele_name_len,
    show_sample_name,
    sample_name_size,
    legend_title_size,
    legend_text_size
  )

}


#' plot_ele_pav
#'
#' Plot the PAV of elements.
#'
#' @param data Coverage results data generated by the APAV tool.
#' @param gff_data GFF data.
#' @param pheno_data Phenotype data. The row names are sample names.
#'
#' @param cluster_samples A logical value indicating whether perform clustering on samples.
#' @param clustering_distance Method of measuring distance when clustering, pass to \code{\link[stats]{dist}}.
#' @param clustering_method Method to perform hierarchical clustering, pass to \code{\link[stats]{hclust}}.
#'
#' @param ele_region Gene element regions for display, choosing from "TN"(transcripts number), "UP"(upstream) and "DOWN"(downstream).
#'
#' @param top_anno_height The relative height of annotation, ranging from 0 to 1.
#' @param left_anno_width The relative width of annotation, ranging from 0 to 1.
#'
#' @param ele_color The color of element regions.
#' @param ele_line_color The color for the line pointing to elements.
#' @param pav_colors A named vector of colors for presence and absence.
#' e.g. c(presence = "#5680ae", absence = "gray70")
#' @param cell_border The color of cell borders in heat map.
#' @param gene_colors A named vector for gene elements.
#' e.g. c(CDS = "#d77478", five_prime_UTR = "#ecbf40")
#' @param pheno_info_color_list A list contains named vector of colors for phenotype.
#' e.g. list(gender = c("Male" = "green", "Female" = "red"), age = c("yellow", "red"))
#' @param pheno_border The color of cell borders in phenotype annotation.
#'
#' @param seq_name_size The size of sequence name.
#' @param loc_name_size The size of coordinate numbers.
#'
#' @param show_ele_name A logical value indicating whether show element names.
#' @param ele_name_size The size of element names.
#' @param ele_name_len The longest length of element names.
#' @param ele_name_rot The rotation of element names.
#'
#' @param show_sample_name A logical value indicating whether show sample names.
#' @param sample_name_size The size of sample names.
#'
#' @param legend_title_size The size of legend title.
#' @param legend_text_size The size of legend item labels.
#'
#' @import data.table
#' @import ggplot2
#' @import patchwork
#' @importFrom ggnewscale new_scale_fill
#'
#' @export
#'
plot_ele_pav <- function(
    data,
    gff_data = NULL,
    pheno_data = NULL,

    cluster_samples = F,
    clustering_distance = "euclidean",
    clustering_method = "complete",

    ele_region = NULL,

    top_anno_height = 0.4,
    left_anno_width = 0.1,

    ele_color = "#A6CEE3",
    ele_line_color = "gray",
    pav_colors = c(presence = "#5680ae", absence = "gray70"),
    cell_border = NA,
    gene_colors = structure(c("gray70", "gray85", "#d77478", "#ecbf40", "#ecbf40"),
                            names = c("transcript", "exon", "CDS", "five_prime_UTR", "three_prime_UTR")),
    pheno_info_color_list = NULL,
    pheno_border = NA,

    seq_name_size = NULL,
    loc_name_size = NULL,

    show_ele_name = T,
    ele_name_size = NULL,
    ele_name_rot = 90,
    ele_name_len = 20,

    show_sample_name = T,
    sample_name_size = NULL,

    legend_title_size = NULL,
    legend_text_size = NULL
  ){

  if(length(grep(";",c(data$Start, data$End))) > 0){
    for(i in 1:nrow(data)){
      pos_m <- merge_block_pos(data[i, "Start"], data[i, "End"])
      data[i, "Start"] <- pos_m[1]
      data[i, "End"] <- pos_m[2]
      data[i, "Length"] <- pos_m[3]
    }
  }

  pav_colors <- get_pav_palette(pav_colors)

  plot_ele(
    "PAV",
    data,
    ele_region,
    gff_data,
    pheno_data,
    cluster_samples,
    clustering_distance,
    clustering_method,
    top_anno_height,
    left_anno_width,
    ele_color,
    ele_line_color,
    pav_colors,
    cell_border,
    gene_colors,
    pheno_info_color_list,
    pheno_border,
    seq_name_size,
    loc_name_size,
    show_ele_name,
    ele_name_size,
    ele_name_rot,
    ele_name_len,
    show_sample_name,
    sample_name_size,
    legend_title_size,
    legend_text_size
  )

}

plot_ele <- function(
  name,
  data,
  ele_region,
  gff_data,
  pheno_data,

  cluster_samples,
  clustering_distance,
  clustering_method,

  top_anno_height,
  left_anno_width,

  ele_color,
  ele_line_color,
  ht_colors,
  cell_border,
  gene_colors,
  pheno_info_color_list,
  pheno_border,

  seq_name_size,
  loc_name_size,

  show_ele_name,
  ele_name_size,
  ele_name_rot,
  ele_name_len,

  show_sample_name,
  sample_name_size,

  legend_title_size,
  legend_text_size
  ){

  loc_breaks = NULL

  match_color("ele_color", "#A6CEE3", ele_color, 1)
  match_color("ele_line_color", "gray", ele_line_color, 1)

  gene_colors <- get_gene_palette(gene_colors)

  if(!is.null(pheno_data)){
    pheno_colors <- get_anno_palette(pheno_info_color_list, as.list(pheno_data))
  }else{
    pheno_colors <- NULL
  }

  top_anno_height <- match_num("top_anno_height", top_anno_height, 0, 1)
  left_anno_width <- match_num("left_anno_width", left_anno_width, 0, 1)

  seqname <- data$Chr[1]

  if(is.numeric(data$Start)){
    data <- data[order(data$Start),]
    ele_range <- c(min(data$Start), max(data$End))
  }else{
    data <- data[order(as.numeric(unlist(lapply(strsplit(data$Start, ";|,"), function(x){x[[1]]})))),]
    ele_range <- c(
      min(as.numeric(unlist(strsplit(data$Start, ";|,")))),
      max(as.numeric(unlist(strsplit(data$End, ";|,"))))
    )
  }

  if(is.null(gff_data)){
    loc_start <- ele_range[1] - 1
    loc_end <- ele_range[2] + 1
  }else{
    colnames(gff_data) <- c("seqid", "source", "type", "start", "end", "score", "strand", "phase", "attr")
    loc_start <- min(ele_range[1], min(gff_data$start)) - 1
    loc_end <-  max(ele_range[2], max(gff_data$end)) + 1
  }

  if(!is.null(ele_region)){
    fdata <- filter_eledata(data, ele_region)
    if(is.null(fdata)){
      stop("Can not find ",ele_region,".\n")
    }
  }else{
    fdata <- data
  }

  p_ele_res <- get_ele(fdata, ele_color, ele_line_color, loc_start, loc_end)
  p_ele <- p_ele_res$p
  enrow <- p_ele_res$nrow + 1

  ps_ht <- get_ht_ps(name,
                     fdata,
                     cluster_samples,
                     clustering_distance,
                     clustering_method,
                     ht_colors,
                     cell_border,
                     pheno_data,
                     pheno_colors,
                     pheno_border,
                     show_ele_name,
                     ele_name_size,
                     ele_name_rot,
                     ele_name_len,
                     show_sample_name,
                     sample_name_size,
                     legend_text_size,
                     legend_title_size)
  p_ht <- ps_ht$ht
  p_anno <- ps_ht$anno

  if(is.null(gff_data)){

    p_ele <- add_coord(p_ele, seqname, loc_start, loc_end, seq_name_size, loc_name_size)

    p_ele + p_ht + p_anno +
      patchwork::plot_layout(design = "
                      #1
                      32
                      ",
                             heights = c(top_anno_height, 1 - top_anno_height),
                             widths= c(left_anno_width, 1 - left_anno_width))

  }else{

    p_struc_res <- get_struc(gff_data, gene_colors)
    p_struc <- p_struc_res$p
    gnrow <- p_struc_res$nrow

    p_struc <- add_coord(p_struc, seqname, loc_start, loc_end, seq_name_size, loc_name_size)

    p_ele <- p_ele +
      scale_x_continuous(expand = expansion(mult=0), limits = c(loc_start, loc_end))

    p_struc + p_ele + p_ht + p_anno +
      patchwork::plot_layout(design = "
                      #1
                      #2
                      43
                      ",
                             heights = c(top_anno_height * (gnrow / (gnrow + enrow)),
                                         top_anno_height * (enrow / (gnrow + enrow)),
                                         (1 - top_anno_height)),
                             widths= c(left_anno_width, 1 - left_anno_width))
  }

}


add_coord <- function(p, seqname, loc_start, loc_end, seq_name_size, loc_name_size){
  loc_breaks <- pretty(range(c(loc_start, loc_end)))
  p <- p +
    labs(x = seqname) +
    scale_x_continuous(expand = expansion(mult=0),
                       limits = c(loc_start, loc_end),
                       breaks = loc_breaks, labels = add_comma(loc_breaks),
                       position = "top") +
    theme(
      axis.text.x = element_text(color = "black", face = "bold", size = loc_name_size),
      axis.title.x = element_text(color = "black", face = "bold", size = seq_name_size),
      axis.line.x = element_line(),
      axis.ticks.x = element_line()
    )
  return(p)
}



get_depth_ps <- function(depth_table,
                         cluster_samples,
                         clustering_distance,
                         clustering_method,
                         log10,
                         depth_colors,
                         pheno_data,
                         pheno_colors,
                         pheno_border,
                         lowlight_colors,
                         ele_pos,
                         show_sample_name,
                         sample_name_size,
                         legend_title,
                         legend_title_size,
                         legend_text_size){

  Pos = Sample = Depth = NULL

  anno_res <- get_anno(depth_table[, 3:ncol(depth_table)],
                       pheno_data, pheno_colors, pheno_border,
                       cluster_samples,
                       clustering_distance, clustering_method,
                       legend_title_size, legend_text_size)
  p_anno <- anno_res$p
  sample_label <- anno_res$label

  depth_pdata <- data.table::melt(data.table::data.table(depth_table),
                                  id.vars = c("Chr", "Pos"), variable.name = "Sample",
                                  value.name = "Depth")

  depth_pdata <- as.data.frame(depth_pdata)
  depth_pdata$Sample <- factor(depth_pdata$Sample, levels = sample_label)
  if(log10){
    depth_pdata$Depth <- log10(depth_pdata$Depth + 1)
  }

  if(is.null(lowlight_colors)){
    p_depth <- ggplot(depth_pdata, aes(x = Pos, y = Sample, fill = Depth)) +
      geom_tile() +
      scale_fill_gradientn(colors = rev(depth_colors))
  }else{
    depth_pdata_region <- depth_pdata[(depth_pdata$Pos %in% ele_pos),]
    depth_pdata_others <- depth_pdata[!(depth_pdata$Pos %in% ele_pos),]

    depth_min <- min(depth_pdata$Depth)
    depth_max <- max(depth_pdata$Depth)

    p_depth <- ggplot() +
      geom_tile(data = depth_pdata_region, aes(x = Pos, y = Sample, fill = Depth)) +
      scale_fill_gradientn(colors = rev(depth_colors), limits = c(depth_min, depth_max)) +
      labs(fill = legend_title) +
      ggnewscale::new_scale_fill() +
      geom_tile(data = depth_pdata_others, aes(x = Pos, y = Sample, fill = Depth)) +
      scale_fill_gradientn(colors = rev(lowlight_colors), limits = c(depth_min, depth_max))
  }

  p_depth <- p_depth +
    scale_y_discrete(expand = expansion(mult=0), position = "right") +
    labs(fill = legend_title) +
    theme(
      axis.line = element_blank(),
      axis.text.x = element_blank(),
      axis.ticks = element_blank(),
      axis.title = element_blank(),
      legend.text = element_text(color = "black", face = "bold", size = legend_text_size),
      legend.title = element_text(color = "black", face = "bold", size = legend_title_size),
      panel.grid = element_blank(),
      panel.background = element_blank(),
      plot.margin = unit(c(2,0,5,0),"mm") ## top right bottom left
    )



  if(show_sample_name){
    p_depth <- p_depth +
      theme(axis.text.y = element_text(color = "black", face="bold",
                                       size = sample_name_size))
  }else{
    p_depth <- p_depth +
      theme(axis.text.y = element_blank())
  }

  return(list(depth = p_depth, anno = p_anno))

}



get_ht_ps <- function(name,
                    data,
                    cluster_samples,
                    clustering_distance,
                    clustering_method,
                    ht_colors,
                    cell_border,
                    pheno_data,
                    pheno_colors,
                    pheno_border,
                    show_ele_name,
                    ele_name_size,
                    ele_name_rot,
                    ele_name_len,
                    show_sample_name,
                    sample_name_size,
                    legend_text_size,
                    legend_title_size){

  Element = Sample = Value = NULL

  cov_d <-  data.frame(apply(data[, 6:ncol(data)], 2, as.numeric), check.names = F)
  if(nrow(data) == 1){
    cov_d <- data.frame(t(cov_d), check.names = F)
  }

  anno_res <- get_anno(cov_d, pheno_data, pheno_colors, pheno_border,
                       cluster_samples,
                       clustering_distance, clustering_method,
                       legend_title_size, legend_text_size)
  p_anno <- anno_res$p
  sample_label <- anno_res$label

  cov_d$Annotation <- data$Annotation
  ele_name <- unique(cov_d$Annotation)
  cov_ld <- data.table::melt(data.table::data.table(cov_d), id.vars = "Annotation")
  colnames(cov_ld) <- c("Element", "Sample", "Value")
  cov_ld <- as.data.frame(cov_ld)
  cov_ld$Sample <- factor(cov_ld$Sample, levels = sample_label)
  cov_ld$Element <- factor(cov_ld$Element, levels = ele_name)

  if(name == "PAV"){
    if(!all(unique(cov_ld$Value) %in% c(0, 1))) stop("The sample columns should contain only 0 and 1.")
    cov_ld$Value <- ifelse(cov_ld$Value == 0, "absence", "presence")
    p_ht <- ggplot(cov_ld, aes(x = Element, y = Sample, fill = Value)) +
      scale_fill_manual(values = ht_colors)
  }else{
    p_ht <- ggplot(cov_ld, aes(x = Element, y = Sample, fill = Value)) +
      scale_fill_gradientn(colors = rev(ht_colors))
  }

  p_ht <- p_ht +
    geom_tile(color = cell_border) +
    scale_x_discrete(expand = expansion(mult=0),
                     breaks = ele_name,
                     labels = trunc_str(ele_name, ele_name_len))+
    scale_y_discrete(expand = expansion(mult=0),
                     position ="right") +
    labs(fill = name) +
    theme(
      axis.ticks = element_blank(),
      axis.title = element_blank(),
      legend.position = "right",
      legend.text = element_text(color = "black", face = "bold", size = legend_text_size),
      legend.title = element_text(color = "black", face = "bold", size = legend_title_size),
      plot.background = element_blank(),
      plot.margin = unit(c(0,0,5,0),"mm") ## top right bottom left
    )

  if(show_ele_name){
    if(ele_name_rot == 0){
      p_ht <- p_ht +
        theme(axis.text.x = element_text(color = "black", face="bold",
                                         size = ele_name_size,
                                         angle = 0))
    }else if (ele_name_rot == 90 | ele_name_rot == 270){
      p_ht <- p_ht +
        theme(axis.text.x = element_text(color = "black", face="bold",
                                         size = ele_name_size,
                                         angle = 90, hjust=1, vjust=.5))
    }else{
      p_ht <- p_ht +
        theme(axis.text.x = element_text(color = "black", face="bold",
                                         size = ele_name_size,
                                         angle = ele_name_rot, hjust = 1))
    }
  }else{
    p_ht <- p_ht +
      theme(axis.text.x = element_blank())
  }

  if(show_sample_name){
    p_ht <- p_ht +
      theme(axis.text.y = element_text(color = "black", face="bold",
                                       size = sample_name_size))
  }else{
    p_ht <- p_ht +
      theme(axis.text.y = element_blank())
  }
  return(list(ht = p_ht, anno = p_anno))
}


get_struc <- function(gff_data, gene_colors){

  type = start = end = tn = NULL

  gff_data$type <- gsub("3UTR", "three_prime_UTR", gff_data$type)
  gff_data$type <- gsub("5UTR", "five_prime_UTR", gff_data$type)

  gff_data$parent <- unlist(lapply(gff_data$attr, function(x){get_attr(x, "Parent")}))
  gff_data$ID <- unlist(lapply(gff_data$attr, function(x){get_attr(x, "ID")}))
  struc_g <- subset(gff_data, type == "gene")
  struc_t <- subset(gff_data, type %in% c("transcript", "mRNA"))
  struc_t$tn <- 1:nrow(struc_t)
  tran_map <- structure(struc_t$tn, names = struc_t$ID)
  struc_e <- subset(gff_data, type == "exon")
  struc_e$tn <- tran_map[struc_e$parent]
  struc_c <- subset(gff_data, type == "CDS")
  struc_c$tn <- tran_map[struc_c$parent]
  struc_u <- subset(gff_data, type %in% c("five_prime_UTR", "three_prime_UTR"))
  struc_u$tn <- tran_map[struc_u$parent]

  struc_width <- 0.8

  p_struc <- ggplot() +
    geom_rect(data = struc_t, aes(xmin = start-.5, xmax = end+.5, ymin = tn-.1*struc_width, ymax= tn+.1*struc_width, fill = type)) +
    geom_rect(data = struc_e, aes(xmin = start-.5, xmax = end+.5, ymin = tn-.5*struc_width, ymax= tn+.5*struc_width, fill = type)) +
    geom_rect(data = struc_c, aes(xmin = start-.5, xmax = end+.5, ymin = tn-.4*struc_width, ymax= tn+.4*struc_width, fill = type)) +
    geom_rect(data = struc_u, aes(xmin = start-.5, xmax = end+.5, ymin = tn-.2*struc_width, ymax= tn+.2*struc_width, fill = type)) +
    theme_classic() +
    scale_fill_manual(values = gene_colors) +
    scale_y_continuous(expand = expansion(mult=c(0,.05)),
                       transform = "reverse") +
    theme(
      axis.title.x = element_blank(),
      axis.title.y = element_blank(),
      axis.text.x = element_blank(),
      axis.text.y = element_blank(),
      axis.ticks.x = element_blank(),
      axis.ticks.y = element_blank(),
      axis.line.x = element_blank(),
      axis.line.y = element_blank(),
      legend.position = "right",
      legend.text = element_text(color = "black", face = "bold"),
      legend.title = element_blank(),
      plot.background = element_blank(),
      plot.margin = unit(c(5,0,0,0),"mm") ## top right bottom left
    )
  gnrow <- nrow(struc_t)
  return(list(p = p_struc, nrow = gnrow))
}


get_ele <- function(data, ele_color, ele_line_color, loc_start, loc_end){

  ele_width <- 0.5

  Start = End = level = xstart = NULL

  region_name <- gsub(":.*","",data$Annotation[1])
  eval(parse(text = paste0("data$Annotation <- gsub('",region_name,":', '', data$Annotation)")))
  ele_res <- get_ele_d(data)
  ele_pd <- ele_res$ele_pd
  ele_pl <- ele_res$ele_pl
  ele_ps <- ele_res$ele_ps

  ele_start <- loc_start
  ele_end <- loc_end
  ele_n <- nrow(ele_ps)
  ele_ps$xstart <- seq(ele_start, ele_end, length.out = (ele_n+1))[1:ele_n] +
    (ele_end - ele_start)/ele_n/2

  p_ele <- ggplot() +
    geom_rect(data = ele_pl, aes(xmin = Start-.5, xmax = End+.5,
                                 ymin = level-.05, ymax = level+.05),
              linewidth = 2, fill = ele_color) +
    geom_rect(data = ele_pd, aes(xmin = Start-.5, xmax = End+.5,
                                 ymin = level - ele_width/2, ymax = level + ele_width/2, fill = "Element regions"),
              colour = "gray50", linewidth = .2) +
    theme_classic() +
    scale_fill_manual(values = c(`Element regions` = ele_color)) +
    scale_y_continuous(expand=expansion(mult=c(0,.1))) +
    theme(
      axis.title.x = element_blank(),
      axis.title.y = element_blank(),
      axis.text.x = element_blank(),
      axis.text.y = element_blank(),
      axis.ticks.x = element_blank(),
      axis.ticks.y = element_blank(),
      axis.line.x = element_blank(),
      axis.line.y = element_blank(),
      plot.background = element_blank(),
      legend.position = "right",
      legend.title = element_blank(),
      legend.text = element_text(color = "black", face = "bold"),
      plot.margin = unit(c(0,0,0,0),"mm") ## top right bottom left
    )

  if(!is.na(ele_line_color)){
    p_ele <- p_ele +
      geom_segment(data = ele_ps, aes(x = (Start + End)/2, y = level - ele_width/2,
                                      xend = xstart, yend = min(ele_pd$level) - 1),
                   color = ele_line_color)
  }
  enrow <- length(unique(ele_pd$level))
  return(list(p = p_ele, nrow = enrow))
}



get_anno <- function(table_data,
                     pheno_data,
                     pheno_colors,
                     pheno_border,
                     cluster_samples,
                     clustering_distance,
                     clustering_method,
                     legend_title_size,
                     legend_text_size){

  table_data <- table_data[,rev(colnames(table_data))]
  samples_n <- ncol(table_data)

  if(cluster_samples){
    dend_data <- ggdendro::dendro_data(stats::as.dendrogram(
      stats::hclust(stats::dist(t(table_data), method = clustering_distance),
                    method = clustering_method),
      type = "rectangle") )

    segment_data <- dend_data$segments
    dend_max <- max(c(segment_data$y, segment_data$yend))
    dend_min <- min(c(segment_data$y, segment_data$yend))
    label_data <- dend_data$labels
  }else{
    label_data <- data.frame(label = colnames(table_data))
  }

  wid <- .9

  x = y = xend = yend = NULL

  p_anno <- ggplot() +
    scale_y_continuous(expand = expansion(mult=0))+
    theme_classic() +
    theme(axis.line = element_blank(),
          axis.text.y = element_blank(),
          axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1,
                                     color = "black", face = "bold"),
          axis.title = element_blank(),
          legend.position = "left",
          legend.text = element_text(color = "black", face = "bold", size = legend_text_size),
          legend.title = element_text(color = "black", face = "bold", size = legend_title_size),
          plot.margin = unit(c(0,0,5,5),"mm"),## top right bottom left
          axis.ticks = element_blank())

  if(cluster_samples){
    p_anno <- p_anno +
    geom_tile(aes(x = 0, y = 1:samples_n, width = 0, height = 1), fill = "white") +
      geom_segment(data = segment_data,
                   aes(x = wid * (dend_min - y) / (dend_max - dend_min) - (1 - wid)/2,
                       y = x,
                       xend = wid *(dend_min - yend) / (dend_max - dend_min) - (1 - wid)/2,
                       yend = xend))
  }

  if(!is.null(pheno_data)){
    if(!(all(label_data$label %in% rownames(pheno_data)))){
      stop("Cannot find all samples in the phenotype table. Please check the rownames of phenotype data.")
    }
    pheno_data <- pheno_data[label_data$label,,drop = F]
    for(j in 1:ncol(pheno_data)){
      if(is.na(pheno_border)){
        pheno_border_text <- "color = NA"
      }else{
        pheno_border_text <- paste0("color = '", pheno_border,"'")
      }
      eval(parse(text = paste0("p_anno <- p_anno +
            geom_tile(data = pheno_data,
            aes(x = ", j -.5, ", y = 1:samples_n,
                          width = ", wid, ", height = 1,
                          fill = ", colnames(pheno_data)[j], "), ", pheno_border_text, ")")))
      if(is.numeric(pheno_data[,j])){
        cur_colors <- attributes(pheno_colors[[colnames(pheno_data)[j]]])$colors
        p_anno <- p_anno +
          scale_fill_gradient(high = cur_colors[1], low = cur_colors[2]) +
          ggnewscale::new_scale_fill()
      } else {
        p_anno <- p_anno +
          scale_fill_manual(values = pheno_colors[[colnames(pheno_data)[j]]]) +
          ggnewscale::new_scale_fill()
      }
    }
    p_anno <- p_anno +
      scale_x_continuous(expand = expansion(mult=0),
                         breaks = (1:ncol(pheno_data)) - .5,
                         labels = colnames(pheno_data))
  } else {
    p_anno <- p_anno +
      scale_x_continuous(expand = expansion(mult=0)) +
      theme(axis.text.x = element_blank())
  }

  return(list(p = p_anno, label = label_data$label))

}

get_ele_d <- function(data){
  ele_d <- data[,1:5]
  ele_d$n <- 1:nrow(ele_d)

  ele_d$Start <- as.character(ele_d$Start)
  ele_d$End <- as.character(ele_d$End)
  ele_d$Length <- as.character(ele_d$Length)

  ele_pd <- data.frame()
  ele_pl <- data.frame()
  ele_ps <- data.frame()
  for(i in 1:nrow(ele_d)){
    cur_s <- as.numeric(unlist(strsplit(ele_d[i, "Start"],",")))
    sl <- length(cur_s)
    cur_e <- as.numeric(unlist(strsplit(ele_d[i, "End"],",")))
    el <- length(cur_e)
    cur_len <- unlist(strsplit(ele_d[i, "Length"],","))
    cur_m <- cur_s + (cur_e - cur_s) /2
    if(i == 1){
      l_e <- cur_e[el]
      level <- 0
    }else{
      if(cur_s[1] <= l_e){
        level = level - 1
      }else{
        level <- 0
      }
      if(cur_e[el] > l_e){
        l_e <- cur_e[el]
      }

    }
    ele_d[i, "level"] <- level
    if(length(cur_s) == 1){
      ele_pd <- rbind(ele_pd, ele_d[i,])
      ele_ps <- rbind(ele_ps, ele_d[i, c("Start", "End", "n", "level")])
    }else{
      ele_pd <- rbind(ele_pd,
                      data.frame(Chr = ele_d[i, "Chr"],
                                 Start = cur_s,
                                 End = cur_e,
                                 Length = cur_len,
                                 Annotation = ele_d[i, "Annotation"],
                                 n = ele_d[i, "n"],
                                 level = ele_d[i, "level"]))
      ele_pl <- rbind(ele_pl,
                      data.frame(Start = cur_m[1],
                                 End = cur_m[length(cur_m)],
                                 n = ele_d[i, "n"],
                                 level = ele_d[i, "level"]))
      ele_ps <- rbind(ele_ps,
                      data.frame(Start = cur_s[1],
                                 End = cur_e[el],
                                 n = ele_d[i, "n"],
                                 level = ele_d[i, "level"]))
    }
  }
  ele_pd$Start <- as.numeric(ele_pd$Start)
  ele_pd$End <- as.numeric(ele_pd$End)
  ele_pl$Start <- as.numeric(ele_pl$Start)
  ele_pl$End <- as.numeric(ele_pl$End)
  ele_ps$Start <- as.numeric(ele_ps$Start)
  ele_ps$End <- as.numeric(ele_ps$End)

  return(list(ele_pd = ele_pd, ele_pl = ele_pl, ele_ps = ele_ps))
}

get_blocks_region <- function(starts, ends){
  all_s <- paste0(starts, collapse = ",")
  all_e <- paste0(ends, collapse = ",")

  pos_m <- merge_block_pos(all_s, all_e)
  rs <- as.numeric(unlist(strsplit(as.character(pos_m[1]), ",")))
  re <- as.numeric(unlist(strsplit(as.character(pos_m[2]), ",")))
  return(unlist(lapply(1:length(rs),
         function(x){
           rs[x]:re[x]
         })))
}


merge_block_pos <- function(str_s, str_e){
  s_arr <- as.numeric(unlist(strsplit(str_s, ",|;")))
  e_arr <- as.numeric(unlist(strsplit(str_e, ",|;")))

  if(length(s_arr) != length(e_arr)){
    stop("Start and end coordinates do not match.")
  }

  if(length(s_arr) == 1){
    return(c(s_arr, e_arr, e_arr - s_arr + 1))
  }else{
    arr_order <- order(s_arr, decreasing = F)
    s_arr <- s_arr[arr_order]
    e_arr <- e_arr[arr_order]

    ts <- s_arr[1]
    te <- e_arr[1]
    tl <- e_arr[1] - s_arr[1] + 1

    fs <- c()
    fe <- c()
    fl <- c()

    for(i in 2:length(arr_order)){
      if(s_arr[i] > te){
        fs <- c(fs, ts)
        fe <- c(fe, te)
        fl <- c(fl, te - ts + 1)
        ts <- s_arr[i]
        te <- e_arr[i]
      }else{
        if(e_arr[i] > te){
          te <- e_arr[i]
        }
      }
    }

    fs <- c(fs, ts)
    fe <- c(fe, te)
    fl <- c(fl, te - ts + 1)

    return(c(paste0(fs, collapse = ","),
             paste0(fe, collapse = ","),
             paste0(fl, collapse = ",")))
  }
}


filter_eledata <- function(eledata, tn){
  #if(!is.numeric(eledata$Start) || !is.numeric(eledata$End)){
  #  stop("Only accepts unmerged element data.")
  #}

  tn_eledata <- do.call(rbind, lapply(1:nrow(eledata), function(i){
    cur_exons <- get_exons(tn, eledata[i, "Annotation"])
    if(cur_exons != ""){
      cur_data <- eledata[i,]
      cur_data$Annotation <- cur_exons
      cur_data
    }else{
      NULL
    }
  }))
  tn_eledata
}


get_exons <- function(tn, anno){
  cur_blocks <- sub("^[^:]+:", "", anno)
  res <- unlist(lapply(unlist(strsplit(cur_blocks, ";")), function(b){
    cur_exons <- gsub("\\]$","",gsub("^\\[", "", b))
    cur_t_exons <- unlist(lapply(unlist(strsplit(cur_exons, ",")), function(x){
      xs <- unlist(strsplit(x, ":"))
      if(xs[1] == tn ){
        x
      }else{
        NULL
      }
    }))
  }))
  paste0(sort(res), collapse = ",")
}


